pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/forward/CLPMutableForwardIndexV2.java

```diff
@@ -0,0 +1,468 @@
+ *   <li>{@code appendEncodedMessage(@NotNull EncodedMessage clpEncodedMessage)}</li>
+ * </ul>
+ *
+ * <p><b>Limitations:</b> The current CLP mutable forward index does not achieve the same compression ratio as the
+ * original standalone CLP implementation, primarily due to design differences between Pinot and CLP. While Pinot
+ * is optimized for fast random access, CLP is designed for single-pass streaming compression and search. As a result,
+ * Pinot sacrifices some compression efficiency for these primary reasons:
+ * <ul>
+ *   <li>Pinot implementation uses block compression compared to CLP’s standalone streaming compression, which achieves
+ *   much better compression ratio at the cost of random access performance.</li>
+ *   <li>Pinot implementation uses uncompressed dictionaries for random lookups of log types and dictionary variables,
+ *   whereas CLP employs compressed dictionaries suited only for single-pass streaming queries.</li>
+ *   <li>Pinot stores additional offsets and length metadata for log types, dictionary variables, and encoded
+ *   variables:</li>
+ *   <ul>
+ *      <li>Streaming forward indices used by CLP's standalone implementation, do not need to store document start or
+ *      end markers because the boundary of one document naturally aligns with the next.</li>
+ *      <li>CLP avoids storing the number of {@code dictVars} and {@code encodedVars}, as this information is already
+ *      embedded in the log type and available during decoding. Pinot, however, needs to store this metadata,
+ *      which can sometimes take up more space than the data itself when compressed.</li>
+ *   </ul>
+ * </ul>
+ * Additionally, CLP standalone binaries can perform search without decompressing the data into plain text. The most
+ * common query type on log data is partial matching. Frequently, searches can be completed by scanning only the logtype
+ * dictionaries for partial matches. Searches can also be executed on {@code logtypeId} and {@code dictVarId} directly
+ * by first performing lookup on the dictionaries to get a subset of dictionary ids to filter, whereas in Pinot,
+ * each dictionary id must be first converted back to strings, followed by a brute-force search on the corresponding
+ * string value. For these reasons, direct searching on CLP columns in Pinot is not yet implemented but may be
+ * included in future updates.</p>
+ */
+public class CLPMutableForwardIndexV2 implements MutableForwardIndex {
+  protected static final Logger LOGGER = LoggerFactory.getLogger(CLPMutableForwardIndexV2.class);
+  public final String _columnName;
+
+  protected final EncodedMessage _clpEncodedMessage;
+  protected final MessageEncoder _clpMessageEncoder;
+  protected final MessageDecoder _clpMessageDecoder;
+
+  protected int _nextDocId = 0;
+  protected int _nextDictVarDocId = 0;
+  protected int _nextEncodedVarId = 0;
+  protected int _bytesRawFwdIndexDocIdStartOffset = Integer.MAX_VALUE;
```

Comment:
is this only for CLPV2? can we add a comment on how this value is dynamic updated?

Reply:
Yes, this is only used for CLPV2 (but might be used for CLPV3 in the future). Internally, we have classes which extends the functionality of this class, so some class member variables are marked as protected.

Reply:
The value is updated at the end of `appendEncodedMessage` method call. I will add a comment to make this clear.